test_runner/regress/test_tenant_size.py

```diff
@@ -0,0 +1,342 @@
+        )
+
+    wait_for_last_flush_lsn(env, first_branch_pg, tenant_id, first_branch_timeline_id)
+    size_after_growing_branch = get_tenant_size(http_client, tenant_id)[0]
+    assert size_after_growing_branch > size_after_branch
+
+    with main_pg.cursor() as cur:
+        cur.execute(
+            f"CREATE TABLE t1 AS SELECT i::bigint n FROM generate_series(0, 2*{batch_size}) s(i)"
+        )
+
+    wait_for_last_flush_lsn(env, main_pg, tenant_id, main_timeline_id)
+    size_after_continuing_on_main = get_tenant_size(http_client, tenant_id)[0]
+    assert size_after_continuing_on_main > size_after_growing_branch
+
+    second_branch_timeline_id = env.neon_cli.create_branch(
+        "second-branch", main_branch_name, tenant_id
+    )
+    size_after_branch = get_tenant_size(http_client, tenant_id)[0]
+    assert size_after_branch > size_after_continuing_on_main
+
+    second_branch_pg = env.postgres.create_start("second-branch", tenant_id=tenant_id)
+
+    with second_branch_pg.cursor() as cur:
+        cur.execute(
+            f"CREATE TABLE t2 AS SELECT i::bigint n FROM generate_series(0, 3*{batch_size}) s(i)"
+        )
+
+    wait_for_last_flush_lsn(env, second_branch_pg, tenant_id, second_branch_timeline_id)
+    size_after_growing_branch = get_tenant_size(http_client, tenant_id)[0]
+    assert size_after_growing_branch > size_after_branch
+
+    with second_branch_pg.cursor() as cur:
+        cur.execute("DROP TABLE t0")
+        cur.execute("DROP TABLE t1")
+        cur.execute("VACUUM FULL")
+
+    wait_for_last_flush_lsn(env, second_branch_pg, tenant_id, second_branch_timeline_id)
+    size_after_thinning_branch = get_tenant_size(http_client, tenant_id)[0]
+    # size doesn't yet go down
+    assert size_after_thinning_branch > size_after_growing_branch
+
+    first_branch_pg.stop_and_destroy()
+    second_branch_pg.stop_and_destroy()
+    main_pg.stop()
+    env.pageserver.stop()
+    env.pageserver.start()
+    size_after = get_tenant_size(http_client, tenant_id)[0]
+
+    assert size_after == size_after_thinning_branch
+
+    # teardown, delete branches, and the size should be going down
+    deleted = False
+    for _ in range(10):
+        try:
+            http_client.timeline_delete(tenant_id, first_branch_timeline_id)
+            deleted = True
+            break
+        except PageserverApiException as e:
+            # compaction is ok but just retry if this fails; related to #2442
+            if "cannot lock compaction critical section" in str(e):
+                time.sleep(1)
+                continue
+            raise
+
+    assert deleted
+
+    size_after_deleting_first = get_tenant_size(http_client, tenant_id)[0]
+    assert size_after_deleting_first < size_after_thinning_branch
+
+    http_client.timeline_delete(tenant_id, second_branch_timeline_id)
+    size_after_deleting_second = get_tenant_size(http_client, tenant_id)[0]
+    assert size_after_deleting_second < size_after_deleting_first
```

Comment:
What would be the "reverted initial" size? Presumably, not 0 anymore, since `gc_cutoff_lsn` is not negative anymore?

Reply:
Added assertions after refactoring not to reuse variables in the test: [90f9f38](https://github.com/neondatabase/neon/pull/2714/commits/90f9f385ad68e0a2df68e8f79287359c4ba55be5)